process: improve nextTick() performance

mscdex opened the pull request:
I recently noticed that with the introduction of Async Hooks also came the (perhaps unintentional) reverting of 804d57db676d54e237a19d42bf443db2b9796525. I actually tried *a lot* of different solutions, but almost all of them resulted in regressions in other `nextTick()` benchmarks (e.g. a perf boost in no args case, but regression in args case) which I still don't quite understand. One thing I did notice is that the ideal solution of just using a plain object and embedding the symbol properties as computed properties causes TurboFan to kick in and that causes a significant slowdown for `nextTick()`.

I was able to improve the existing non-ES6-based constructor solution, but it turns out using an ES6 class improves performance even more (even surpassing the plain object-based solutions that performed better in whichever benchmarks), so that is what I decided on.

Results:
```
                                               improvement confidence      p.value
 process/next-tick-breadth-args.js millions=2     27.75 %        *** 1.271176e-20
 process/next-tick-breadth.js millions=2           7.71 %        *** 4.155765e-13
 process/next-tick-depth-args.js millions=12      47.78 %        *** 4.150674e-52
 process/next-tick-depth.js millions=12           47.32 %        *** 7.742778e-31
```

CI: https://ci.nodejs.org/job/node-test-pull-request/8464/

##### Checklist

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)

* process

refack commented on lib/internal/process/next_tick.js:
```diff
@@ -210,43 +214,41 @@ function setupNextTick() {
+      this.domain = process.domain || null;
+      this[async_id_symbol] = asyncId;
+      this[trigger_id_symbol] = triggerId;
+      if (async_hook_fields[kInit] > 0)
```
I'm not 100% with so much side effects in a constructor...
If it's not a performance hit I'd rather see these four lines in a static method or factory
If it is a performance hit, I think it should be documented:
```
// These side effects were folded into the c'tor for performance
```
(or something to that effect)

mscdex replied:
You can't move these other statements back into `nextTick()` without causing a performance regression from what I saw. I believe it's because of Crankshaft's inlining limits, but I did not want to introduce regressions at this moment (we will have to deal with it once TurboFan is everywhere).

refack replied:
Yeah so probably even a comment will "irritate" Crankshaft
What can you do 🤷‍♂️

mscdex replied:
Hopefully once V8 (or perhaps they already do in later versions) is able to better optimize computed properties in TurboFan, we can use plain objects again without losing performance.

mscdex commented on lib/internal/process/next_tick.js:
```diff
@@ -86,8 +89,9 @@ function setupNextTick() {
 
     if (tickInfo[kIndex] < tickInfo[kLength] ||
-        emitPendingUnhandledRejections())
+        emitPendingUnhandledRejections()) {
```
Style choice, I think we tend to use braces for multi-line conditionals from what I've generally seen, this is just one "offender" I ran across while working in this file.

Fishrock123 commented on lib/internal/process/next_tick.js:
```diff
@@ -210,43 +214,41 @@ function setupNextTick() {
     if (typeof callback !== 'function')
       throw new errors.TypeError('ERR_INVALID_CALLBACK');
-    // on the way out, don't bother. it won't get fired anyway.
+
```
Can we clarify the comment rather than removing it?

mscdex replied:
Removing it matches `internalNextTick()` and it's not a particularly helpful comment as `process._exiting` is pretty self-explanatory to me.

Fishrock123 replied:
Yes, the code says _what_ but I'm wondering if it may be helpful to say _why_.

refack replied:
FWIW `Crankshaft` can decide not to inline a function _because_ it has a comment in it 😞 so if we come up with a better comment it probably should go in line 231.

Fishrock123 replied:
Eh, it makes little difference as the comment was already there and that will soon be an obsolete optimization as I understand.

mscdex replied:
In addition to my above response, it is also to do with inlining, as I don't want to submit a PR that degrades performance.

Fishrock123 replied:
¯\\\_(ツ)\_/¯

mscdex replied:
@Fishrock123 I don't understand what you mean by a 'why comment.'

gibfahn replied:
>I don't understand what you mean by a 'why comment.'

A comment explaining why we're doing this (as opposed to explaining what this code does).

mscdex replied:
@gibfahn I understand that part. What I meant was I don't understand what this is in reference to. Is it the `process._exiting` line? The changes I'm making? Something else? What is the suggested 'why comment' ?

gibfahn replied:
> What is the suggested 'why comment' ?

(AIUI) This is the comment, it's a "why comment", it's helpful to preserve it. 

```js
    // on the way out, don't bother. it won't get fired anyway.
```

FWIW I'm sure putting it above the function is fine.

mscdex replied:
@Fishrock123 @gibfahn I've added something above both functions now, let me know if that is what you had in mind or not.

gibfahn replied:
>@Fishrock123 @gibfahn I've added something above both functions now, let me know if that is what you had in mind or not.

Clearer than the original, thanks.

Fishrock123 commented on lib/internal/process/next_tick.js:
```diff
@@ -210,43 +214,41 @@ function setupNextTick() {
-               tickObject);
+  class TickObject {
+    constructor(callback, args, triggerId) {
+      var asyncId = ++async_uid_fields[kAsyncUidCntr];
```
Is there a reason this is  `var` over `const`?

mscdex replied:
I think it was trigger a deopt.

refack commented on lib/internal/process/next_tick.js:
```diff
@@ -210,43 +214,41 @@ function setupNextTick() {
+      tickInfo[kLength]++;
     }
   }
 
```
// we bail fast `if (process._exiting)` since the loop has effectively stopped
// and there will never be a `nextTick`

refack replied:
@Fishrock123 PTAL

AndreasMadsen commented on benchmark/process/next-tick-depth-args.js:
```diff
@@ -2,7 +2,7 @@
 var common = require('../common.js');
 var bench = common.createBenchmark(main, {
-  millions: [2]
+  millions: [12]
```
Why this increase? You are getting ridiculous statistical confidence. By some quick calculation, this should only improve the confidence by a factor of 2.5. Compensating for that and the previous version should also give you confidence well beyond `***`.

mscdex replied:
I always choose something large enough to allow for any re-opts to occur and for results to stabilize. With `2`, I was getting results that varied quite a bit in comparison.

mscdex commented:
FWIW with V8 5.9 in master I now get these results with this PR as-is:

```
                                               improvement confidence      p.value
 process/next-tick-breadth-args.js millions=2      6.17 %        *** 2.725997e-04
 process/next-tick-breadth.js millions=2           6.68 %         ** 1.568356e-03
 process/next-tick-depth-args.js millions=12      46.71 %        *** 2.458707e-81
 process/next-tick-depth.js millions=12           43.66 %        *** 3.338435e-66
```

mscdex commented:
Ok, I've made a couple more adjustments:

* Removed the non-assignment code out of the `TickObject` class constructor as the difference in code location no longer makes any difference with TurboFan in V8 5.9
* Switched from an array to a linked list for storing `TickObject` instances, this increases performance more significantly than the original changes alone. I *think* it may be possible to make some further optimizations to avoid use of the `tickInfo` counters shared with C++ land (at least the `kIndex` counter), however I've decided to leave it alone for now.

Here are the results as of these latest changes:

```
                                               improvement confidence      p.value
 process/next-tick-breadth-args.js millions=4    135.31 %        *** 3.011128e-73
 process/next-tick-breadth.js millions=4         361.40 %        *** 5.097232e-76
 process/next-tick-depth-args.js millions=12      44.08 %        *** 1.276617e-66
 process/next-tick-depth.js millions=12           43.85 %        *** 1.333610e-49
```

mscdex commented:
@refack @Fishrock123 @jasnell @evanlucas Please review these recent changes.

TimothyGu commented on lib/internal/process/next_tick.js:
```diff
@@ -210,45 +251,48 @@ function setupNextTick() {
+      default:
+        args = new Array(arguments.length - 1);
+        for (var i = 1; i < arguments.length; i++)
+          args[i - 1] = arguments[i];
```
This is faster than spread parameters even with V8 5.9?

mscdex replied:
I'm not sure how spread could be used here. Do you mean rest parameters? If so, I tried that and while it's a *little* faster in one case, it's slower to varying degrees in all others.

TimothyGu replied:
Yeah, that's what I meant. Sorry, could never get all the terminologies right.

refack commented:
So I'm just writing down what I saw (writing down mostly to organize my thoughts):
1. the linked list queue
2. inlined `setupInit` X 2
3. `switch` for args X 2
4. the `const microTasksTickObject`

Anything else of significance?

mscdex commented:
@refack That's mostly it, unless you want to also count the switch to ES6 class for `TickObject`.

refack commented:
Thanks for the confirmation ✔️

mscdex commented:
Rebased. If there are no objections in the next 24 hours I will assume all who approved the original changes also approve these latest changes that I previously outlined.
